File: NLPWebService/src/weather/api/WeatherService.java
package weather.api;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

public class WeatherService {

	private static Random rand = new Random();
	public static String URL = "http://wthrcdn.etouch.cn/WeatherApi?city=";
	
	public static String getWeatherByCityName(String city, Date fromDate) {
		return getWeatherByCityName(city, fromDate, fromDate);
	}

	public static String getWeatherByCityName(String city, Date fromDate, Date endDate) {
		String queryUrl = URL + city;
		String status = "";
		HttpClient client = HttpClients.createDefault();
		HttpGet cmd = new HttpGet(queryUrl);
		try {
			HttpResponse res = client.execute(cmd);
			HttpEntity entity = res.getEntity();
			if (entity != null) {
				String restring = EntityUtils.toString(entity,
						Charset.forName("utf-8"));
				if (restring != null && restring.length() != 0) {
					SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd");
					status = digestWeatherInfo(restring, fromDate, endDate, df.format(fromDate).equals(df.format(endDate)) && df.format(fromDate).equals(df.format(new Date())));
				}
			}
		} catch (ClientProtocolException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		return status;
	}

	private static String digestWeatherInfo(String restring, Date fromDate, Date endDate, boolean details) {
		StringBuffer sb = new StringBuffer();
		try {
			DocumentBuilderFactory factory = DocumentBuilderFactory
					.newInstance();
			DocumentBuilder builder = factory.newDocumentBuilder();
			InputStream is = new ByteArrayInputStream(
					restring.getBytes("utf-8"));
			Document document = builder.parse(is);
			Element root = document.getDocumentElement();
			XPath xpath = XPathFactory.newInstance().newXPath();

			long gap = (endDate.getTime() - fromDate.getTime()) / (24 * 3600000);
			long gap2 = gap;
			String highLowTemp = "%1$s%2$s,%3$s,";
			//String day = new SimpleDateFormat("M月dd日").format(fromDate);

			String high = "";
			String low = "";
			String dayInfo = "";
			String nightInfo = "";
			String city = "";

			String datePattern = new SimpleDateFormat("d").format(fromDate) + "日";
			NodeList weathers = (NodeList) xpath.evaluate(
					"/resp/forecast/weather", document, XPathConstants.NODESET);

			city = (String) xpath.evaluate("/resp/city/text()",
					document, XPathConstants.STRING);
			sb.append(city);
			boolean start = false;
			for (int i = 0; i < weathers.getLength(); i++) {
				Node node = weathers.item(i);
				String weatherDay = (String) xpath.evaluate("date/text()",
						node, XPathConstants.STRING);
				if (weatherDay.startsWith(datePattern) || (start && gap >= 0)) {
					start = true;
					high = (String) xpath.evaluate("high/text()", node,
							XPathConstants.STRING);
					high = high.replaceAll("高温", "最高温度");
					low = (String) xpath.evaluate("low/text()", node,
							XPathConstants.STRING);
					low = low.replaceAll("低温", "最低温度");
					if (gap2 == 0) {
						weatherDay = getDayDesc(fromDate);
					} else {
						Date d = new Date();
						d.setTime(fromDate.getTime() + 24*3600000*(gap2-gap));
						weatherDay = new SimpleDateFormat("M月d日").format(d);
					}
					String info = String.format(highLowTemp, new Object[]{weatherDay, high, low});
					sb.append(info);
					
					String dtype = (String) xpath.evaluate("ay/type/text()", node,
							XPathConstants.STRING);
					if (dtype == null || dtype.length() == 0) {
						dtype = "晴";
					}
					String dinfo = "白天" + dtype + ",";
					String wind = (String) xpath.evaluate("day/fengxiang/text()", node,
							XPathConstants.STRING);
					if (wind != null && wind.length() != 0 && wind.indexOf("无") == -1) {
						dinfo += wind + ",";
						String windStrongness = (String) xpath.evaluate("day/fengli/text()", node,
								XPathConstants.STRING);
						if (windStrongness != null && windStrongness.length() != 0) {
							dinfo += windStrongness + ",";
						}
						
					}
					sb.append(dinfo);
					
					
					String ntype = (String) xpath.evaluate("night/type/text()", node,
							XPathConstants.STRING);
					if (ntype == null || ntype.length() == 0) {
						ntype = "晴";
					}
					String ninfo = "晚上" + ntype + ",";
					
					wind = (String) xpath.evaluate("night/fengxiang/text()", node,
							XPathConstants.STRING);
					if (wind != null && wind.length() != 0 && wind.indexOf("无") == -1) {
						ninfo += wind + ",";
						String windStrongness = (String) xpath.evaluate("night/fengli/text()", node,
								XPathConstants.STRING);
						if (windStrongness != null && windStrongness.length() != 0) {
							ninfo += windStrongness + ",";
						}
						
					}
					sb.append(ninfo);
					
					gap--;
				}
			}
			
			if (details) {
				String envPattern = "今天PM2.5指数为%1$s,属于%2$s,%3$s建议%4$s,";
				String pm25 = (String) xpath.evaluate("/resp/environment/pm25/text()",
						document, XPathConstants.STRING);
				String quality = (String) xpath.evaluate("/resp/environment/quality/text()",
						document, XPathConstants.STRING);
				String major = (String) xpath.evaluate("/resp/environment/MajorPollutants/text()",
						document, XPathConstants.STRING);
				if (major != null && major.length() > 0) {
					major = "主要污染物为" + major + ",";
				}
				String suggest = (String) xpath.evaluate("/resp/environment/suggest/text()",
						document, XPathConstants.STRING);
				sb.append(String.format(envPattern, new Object[]{pm25, quality, major, suggest}));
				
				String zhiShuPattern = "今天%1$s为,%2$s,%3$s";
				NodeList zhishuList = (NodeList) xpath.evaluate(
						"/resp/zhishus/zhishu", document, XPathConstants.NODESET);
				if (zhishuList.getLength() > 0) {
					int ind = rand.nextInt(zhishuList.getLength());
					Node zhishu = zhishuList.item(ind);
					String zhiShuName = (String) xpath.evaluate("name/text()",
							zhishu, XPathConstants.STRING);
					String zhishuValue = (String) xpath.evaluate("value/text()",
							zhishu, XPathConstants.STRING);
					String detail = (String) xpath.evaluate("detail/text()",
							zhishu, XPathConstants.STRING);
					sb.append(String.format(zhiShuPattern, new Object[]{zhiShuName, zhishuValue, detail}));
				}
			}
			
		} catch (ParserConfigurationException | SAXException | IOException e) {
			e.printStackTrace();
		} catch (XPathExpressionException e) {
			e.printStackTrace();
		}

		return sb.toString();
	}

	private static String getDayDesc(Date fromDate) {
		if (fromDate == null) {
			return "";
		}
		
		try {
			SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd");
			Date s = df.parse(df.format(fromDate));
			Date e = df.parse(df.format(new Date()));
			
			int gap = (int) ((s.getTime() - e.getTime()) / (24*3600000));
			switch (gap) {
			case 0:
				return "今天";
			case 1:
				return "明天";
			case 2:
				return "后天";
			}
		} catch (ParseException e) {
			e.printStackTrace();
		}
		
		return new SimpleDateFormat("M月d日").format(fromDate);
	}

	public static void main(String[] args) {
		String time = "北京1月21日天气";
		Pattern p = Pattern.compile("([0-9]{1,2})月([0-9]{1,2})[号|日]");
		Matcher m = p.matcher(time);
		if (m.find()) {
			int gc = m.groupCount();
			for (int i = 0; i <= gc; i++) {
				System.out.println(m.group(i));
			}
		}
		System.out.println(m.find());
//		System.out.println(Pattern.matches("[0-9]{1,2}月[0-9]{1,2}[号|日]", time));
//		System.out.println(Pattern.matches("北京1月21日天气", time));
//		Date s = new Date();
//		Date e = new Date(s.getTime() + 24*3600000);
//		System.out.println(WeatherService.getWeatherByCityName("北京", e, e));
	}
}

File: Speech/NL2ML/utils/WeatherService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Web;
using System.Xml;

namespace NL2ML.utils
{
    public class WeatherService
    {
        private static Random rand = new Random();
        public static string URL = "http://wthrcdn.etouch.cn/WeatherApi?city={0}";

        public static string GetWeatherByCityName(string city, DateTime fromDate)
        {
            return GetWeatherByCityName(city, fromDate, fromDate);
        }

        public static string GetWeatherByCityName(string city, DateTime fromDate, DateTime endDate)
        {
            string queryUrl = URL + city;
            string status = "";
            WebRequest request = WebRequest.Create(string.Format(URL,
                HttpUtility.UrlEncode(city, Encoding.UTF8)));
            // Get the response.
            using (WebResponse response = request.GetResponse())
            {
                // Display the status.
                // Get the stream containing content returned by the server.
                Stream dataStream = response.GetResponseStream();
                using (var zipStream =
                          new System.IO.Compression.GZipStream(dataStream, System.IO.Compression.CompressionMode.Decompress))
                {
                    // Open the stream using a StreamReader for easy access.
                    using (StreamReader reader = new StreamReader(zipStream, Encoding.UTF8))
                    {
                        // Read the content.
                        string restring = reader.ReadToEnd();
                        status = DigestWeatherInfo(restring, fromDate, endDate, fromDate.ToShortDateString().Equals(endDate.ToShortDateString())
                            && fromDate.ToShortDateString().Equals(DateTime.Now.ToShortDateString()));
                    }
                }
                
            }

            return status;
        }

        private static string DigestWeatherInfo(string restring, DateTime fromDate, DateTime endDate, bool details)
        {
            StringBuilder sb = new StringBuilder();

            XmlDocument doc = new XmlDocument();
            doc.LoadXml(restring);
            XmlNode root = doc.DocumentElement;

            long gap = (endDate.DayOfYear - fromDate.DayOfYear);
            long gap2 = gap;
            string highLowTemp = "{0}{1},{2},";
            //string day = new SimpleDateFormat("M月dd日").format(fromDate);

            string high = "";
            string low = "";
            string dayInfo = "";
            string nightInfo = "";
            string city = "";

            string datePattern = fromDate.Day + "日";

            XmlNodeList weathers = root.SelectNodes("//forecast/weather");
            city = root.SelectSingleNode("//city").InnerText;
            sb.Append(city);
            bool start = false;
            for (int i = 0; i < weathers.Count; i++)
            {
                XmlNode node = weathers[i];
                string weatherDay = node.SelectSingleNode("descendant::date").InnerText;//(string) xpath.evaluate("date/text()",node, XPathConstants.STRING);

                if (weatherDay.StartsWith(datePattern) || (start && gap >= 0))
                {
                    start = true;
                    high = node.SelectSingleNode("descendant::high").InnerText;
                    high = high.Replace("高温", "最高温度");
                    low = node.SelectSingleNode("descendant::low").InnerText;
                    low = low.Replace("低温", "最低温度");
                    if (gap2 == 0)
                    {
                        weatherDay = GetDayDesc(fromDate);
                    }
                    else
                    {
                        DateTime d = new DateTime(fromDate.Year, fromDate.Month, fromDate.Day + (int)(gap2 - gap));
                        weatherDay = d.ToString("M月d日");
                    }
                    string info = string.Format(highLowTemp, weatherDay, high, low);
                    sb.Append(info);

                    string dtype = node.SelectSingleNode("descendant::day/type").InnerText;//(string) xpath.evaluate("ay/type/text()", node,XPathConstants.STRING);
                    if (dtype == null || dtype.Length == 0)
                    {
                        dtype = "晴";
                    }
                    string dinfo = "白天" + dtype + ",";
                    string wind = node.SelectSingleNode("descendant::day/fengxiang").InnerText;//(string) xpath.evaluate("day/fengxiang/text()", node,XPathConstants.STRING);
                    if (wind != null && wind.Length != 0 && wind.IndexOf("无") == -1)
                    {
                        dinfo += wind + ",";
                        string windStrongness = node.SelectSingleNode("descendant::day/fengli").InnerText;//(string) xpath.evaluate("day/fengli/text()", node,XPathConstants.STRING);
                        if (windStrongness != null && windStrongness.Length != 0)
                        {
                            dinfo += windStrongness + ",";
                        }

                    }
                    sb.Append(dinfo);


                    string ntype = node.SelectSingleNode("descendant::night/type").InnerText;//(string) xpath.evaluate("night/type/text()", node,XPathConstants.STRING);
                    if (ntype == null || ntype.Length == 0)
                    {
                        ntype = "晴";
                    }
                    string ninfo = "晚上" + ntype + ",";

                    wind = node.SelectSingleNode("descendant::night/fengxiang").InnerText;//(string) xpath.evaluate("night/fengxiang/text()", node,XPathConstants.STRING);
                    if (wind != null && wind.Length != 0 && wind.IndexOf("无") == -1)
                    {
                        ninfo += wind + ",";
                        string windStrongness = node.SelectSingleNode("descendant::night/fengli").InnerText;//(string) xpath.evaluate("night/fengli/text()", node,XPathConstants.STRING);
                        if (windStrongness != null && windStrongness.Length != 0)
                        {
                            ninfo += windStrongness + ",";
                        }

                    }
                    sb.Append(ninfo);

                    gap--;
                }
            }

            if (details)
            {
                string envPattern = "今天PM2.5指数为{0},属于{1},{2}建议{3},";
                string pm25 = root.SelectSingleNode("//environment/pm25").InnerText;//(string) xpath.evaluate("/resp/environment/pm25/text()",document, XPathConstants.STRING);
                string quality = root.SelectSingleNode("//environment/quality").InnerText; //(string)xpath.evaluate("/resp/environment/quality/text()", document, XPathConstants.STRING);
                string major = root.SelectSingleNode("//environment/MajorPollutants").InnerText; //(string)xpath.evaluate("/resp/environment/MajorPollutants/text()", document, XPathConstants.STRING);
                if (major != null && major.Length > 0)
                {
                    major = "主要污染物为" + major + ",";
                }
                string suggest = root.SelectSingleNode("//environment/suggest").InnerText;//(string) xpath.evaluate("/resp/environment/suggest/text()",document, XPathConstants.STRING);
                sb.Append(string.Format(envPattern, pm25, quality, major, suggest));

                string zhiShuPattern = "今天{0}为,{1},{2}";
                XmlNodeList zhishuList = root.SelectNodes("//zhishus/zhishu");//(NodeList) xpath.evaluate("/resp/zhishus/zhishu", document, XPathConstants.NODESET);
                if (zhishuList.Count > 0)
                {
                    int ind = rand.Next(zhishuList.Count);
                    XmlNode zhishu = zhishuList[ind];
                    string zhiShuName = zhishu.SelectSingleNode("descendant::name").InnerText;//(string) xpath.evaluate("name/text()",zhishu, XPathConstants.STRING);
                    string zhishuValue = zhishu.SelectSingleNode("descendant::value").InnerText;//(string) xpath.evaluate("value/text()",zhishu, XPathConstants.STRING);
                    string detail = zhishu.SelectSingleNode("descendant::detail").InnerText;//(string) xpath.evaluate("detail/text()",zhishu, XPathConstants.STRING);
                    sb.Append(string.Format(zhiShuPattern, zhiShuName, zhishuValue, detail));
                }
            }


            return sb.ToString();
        }

        private static string GetDayDesc(DateTime fromDate)
        {
            if (fromDate == null)
            {
                return "";
            }
            int gap = fromDate.DayOfYear - DateTime.Now.DayOfYear;
            switch (gap)
            {
                case 0:
                    return "今天";
                case 1:
                    return "明天";
                case 2:
                    return "后天";
            }

            return fromDate.ToString("M月d日");
        }
    }
}
